src/st/st-widget.c

```diff
@@ -486,8 +500,15 @@ st_widget_style_changed (StWidget *widget)
     }
 
   /* update the style only if we are mapped */
-  if (clutter_actor_is_mapped (CLUTTER_ACTOR (widget)))
-    st_widget_recompute_style (widget, old_theme_node);
+  if (clutter_actor_is_mapped (CLUTTER_ACTOR (widget))
+  || (old_theme_node != NULL &&
+      (old_theme_node->background_blur > 0
+       || old_theme_node->background_bumpmap != NULL)))
+    {
+      st_widget_recompute_style (widget, old_theme_node);
+
+      st_widget_add_background_effects(widget, old_theme_node);
```

Comment:
Maybe we can check if the theme node is eligible before calling `st_widget_add_background_effects`? We should know after `st_widget_recompute_style`.

Reply:
I think that will just move the tests in the function to be done preceding it.

Reply:
After #8230 is merged, you could piggy-back off of the change condition I added, to give some idea. I find it hard to believe you can say there will be no improvement on something you haven't tried yet.